File: core0/us/wallet/trader/data_t.cpp
//===-                           P L E B B L E
//===-                         https://plebble.us
//===-
//===-              Copyright (C) 2017-2022 root1m3@plebble.us
//===-
//===-                      GNU GENERAL PUBLIC LICENSE
//===-                       Version 3, 29 June 2007
//===-
//===-    This program is free software: you can redistribute it and/or modify
//===-    it under the terms of the AGPLv3 License as published by the Free
//===-    Software Foundation.
//===-
//===-    This program is distributed in the hope that it will be useful,
//===-    but WITHOUT ANY WARRANTY; without even the implied warranty of
//===-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//===-
//===-    You should have received a copy of the General Public License
//===-    along with this program, see LICENCE file.
//===-    see https://www.gnu.org/licenses
//===-
//===----------------------------------------------------------------------------
//===-
#include "data_t.h"

#include <us/gov/io/cfg0.h>
#include <us/gov/crypto/base58.h>

#include "trader_t.h"

#define loglevel "wallet/trader"
#define logclass "data_t"
#include <us/gov/logs.inc>

using namespace us::wallet::trader;
using c = us::wallet::trader::data_t;

c::data_t() {
}

c::~data_t() {
    log("destructor", size(), this);
}

c::data_t(const data_t& other): b(other) {
}

c::data_t(data_t&& other): b(move(other)) {
}


template<> void c::add(const string& key, const string& value) {
    emplace(key, value);
}

bool c::operator == (const data_t& other) const {
    if (size() != other.size()) return false;
    auto i = begin();
    auto j = other.begin();
    while (i != end()) {
        if (i->first != j->first) return false;
        if (i->second != j->second) return false;
        ++i;
        ++j;
    }
    return true;
}

bool c::operator != (const data_t& other) const {
    if (size() != other.size()) return true;
    auto i = begin();
    auto j = other.begin();
    while (i != end()) {
        if (i->first != j->first) return true;
        if (i->second != j->second) return true;
        ++i;
        ++j;
    }
    return false;
}

void c::dump(const string& pfx, ostream& os) const {
    lock_guard<mutex> lock(mx);
    dump_(pfx, os);
}

void c::dump_(const string& pfx, ostream& os) const {
    for (auto& i: *this) {
        os << pfx << "'" << i.first << "': " << i.second << '\n';
    }
}

c::diff_t c::get_diff(const data_t& prev) const {
    log("get_diff");
    auto newer = begin();
    auto older = prev.begin();
    diff_t d;
    while (newer != end() && older != prev.end()) {
        if (newer->first == older->first) {
            if (newer->second != older->second) {
                d.emplace(newer->first, diff_line{'M', newer->second});
            }
            ++newer;
            ++older;
            continue;
        }
        if (newer->first < older->first) {
            d.emplace(newer->first, diff_line{'A', newer->second});
            ++newer;
            continue;
        }
        d.emplace(older->first, diff_line{'D', older->second});
        ++older;
    }
    if (!empty()) {
        while (newer != end()) {
            d.emplace(newer->first, diff_line{'A', newer->second});
            ++newer;
        }
    }
    if (!prev.empty()) {
        while (older != prev.end()) {
            d.emplace(older->first, diff_line{'D', older->second});
            ++older;
        }
    }
    log("get_diff returning", d.size(), "lines");
    return move(d);
}

void c::diff_t::dump(const string& pfx, ostream& os) const {
    for (auto& i: *this) {
        os << pfx << i.second.action << ' ' << i.first << ' ' << i.second.value << '\n';
    }
}

void c::diff_t::filter(const string& k) {
    auto i = find(k);
    if (i != end()) {
        erase(i);
    }
}

ko c::apply(const diff_t& d) {
    for (auto& i: d) {
        switch (i.second.action) {
            case 'A': {
                if (!emplace(i.first, i.second.value).second) {
                    auto r = "KO 79658 item was already there.";
                    log(r);
                    return r;
                }
            }
            break;
            case 'M': {
                auto x = find(i.first);
                if (x == end()) {
                    auto r = "KO 79659 item was not found there.";
                    log(r);
                    return r;
                }
                x->second = i.second.value;
            }
            break;
            case 'D': {
                auto x = find(i.first);
                if (x == end()) {
                    auto r = "KO 79660 item was not found there.";
                    log(r);
                    return r;
                }
                erase(x);
            }
            break;
            default: {
                auto r = "KO 79661 unexpected action.";
                log(r);
                return r;
            }
        }
    }
    return ok;
}

ko c::from(const string& payload) {
    istringstream is(payload);
    return from(is);
}

ko c::from(istream& is) {
    log("data_t::from_stream");
    lock_guard<mutex> lock(mx);
    clear();
    while (is.good()) {
        string line;
        getline(is, line);
        if (line.empty()) continue;
        istringstream i2(line);
        string k;
        string v;
        i2 >> k;
        getline(i2, v);
        if (k.empty()) {
            auto r = "KO 66058";
            log(r);
            clear();
            return r;
        }
        if (is.fail()) {
            auto r = "KO 66059";
            log(r);
            clear();
            return r;
        }
        us::gov::io::cfg0::trim(v);
        emplace(k, v);
    }
    return ok;
}

c& c::operator = (const c& other) {
    b::operator = (other);
    return *this;
}

File: core0/us/sdk/wallet/java/us/wallet/trader/data_t.java
//===-                           P L E B B L E
//===-                         https://plebble.us
//===-
//===-              Copyright (C) 2017-2022 root1m3@plebble.us
//===-
//===-                      GNU GENERAL PUBLIC LICENSE
//===-                       Version 3, 29 June 2007
//===-
//===-    This program is free software: you can redistribute it and/or modify
//===-    it under the terms of the AGPLv3 License as published by the Free
//===-    Software Foundation.
//===-
//===-    This program is distributed in the hope that it will be useful,
//===-    but WITHOUT ANY WARRANTY; without even the implied warranty of
//===-    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//===-
//===-    You should have received a copy of the General Public License
//===-    along with this program, see LICENCE file.
//===-    see https://www.gnu.org/licenses
//===-
//===----------------------------------------------------------------------------
//===-
package us.wallet.trader;
import java.io.BufferedReader;                                                                 // BufferedReader
import java.io.ByteArrayInputStream;                                                           // ByteArrayInputStream
import java.io.ByteArrayOutputStream;                                                          // ByteArrayOutputStream
import us.CFG;                                                                                 // CFG
import java.util.HashMap;                                                                      // HashMap
import us.stdint.*;                                                                            // *
import java.io.InputStream;                                                                    // InputStream
import java.io.InputStreamReader;                                                              // InputStreamReader
import static us.ko.is_ko;                                                                     // is_ko
import us.ko;                                                                                  // ko
import java.util.Map;                                                                          // Map
import static us.ko.ok;                                                                        // ok
import java.io.OutputStream;                                                                   // OutputStream
import us.pair;                                                                                // pair
import java.io.PrintStream;                                                                    // PrintStream
import java.util.Scanner;                                                                      // Scanner
import java.util.TreeMap;                                                                      // TreeMap

@SuppressWarnings("serial")
public class data_t extends TreeMap<String, String> {

    static void log(final String line) {           //--strip
       CFG.log_wallet_trader("data_t: " + line);   //--strip
    }                                              //--strip

    public String find(String key) {
        return get(key);
    }

    public void emplace(String k, String v) {
        put(k, v);
    }

    public ko from(String s) {
        log("from string: " + s); //--strip
        InputStream is = new ByteArrayInputStream(s.getBytes());
        return from(is);
    }

    public ko from(InputStream is) {
        log("from stream"); //--strip
        clear();
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(is))) {
            while (reader.ready()) {
                String line = reader.readLine();
                if (line.isEmpty()) continue;
                String[] kv = line.split(" ", 2);
                if (kv.length != 2) return new ko("KO 50045");
                if (kv[0].isEmpty()) {
                    clear();
                    return new ko("KO 66058");
                }
                put(kv[0], kv[1]);
            }
        }
        catch (Exception e) {
        }
        log("size " + size()); //--strip
        return ok;
    }

    public void dump(final String pfx, PrintStream os) {
        dump_(pfx, os);
    }

    public void dump_(final String pfx, PrintStream os) {
        for (Map.Entry<String, String> entry: entrySet()) {
            os.println(pfx + "'" + entry.getKey() + "': " + entry.getValue());
        }
    }

}

